File: ch4/py/e-filter-messages.py
from langchain_core.messages import (
    AIMessage,
    HumanMessage,
    SystemMessage,
    filter_messages,
)

# Sample messages
messages = [
    SystemMessage(content="you are a good assistant", id="1"),
    HumanMessage(content="example input", id="2", name="example_user"),
    AIMessage(content="example output", id="3", name="example_assistant"),
    HumanMessage(content="real input", id="4", name="bob"),
    AIMessage(content="real output", id="5", name="alice"),
]

# Filter for human messages
human_messages = filter_messages(messages, include_types="human")
print("Human messages:", human_messages)

# Filter to exclude certain names
excluded_names = filter_messages(
    messages, exclude_names=["example_user", "example_assistant"]
)
print("\nExcluding example names:", excluded_names)

# Filter by types and IDs
filtered_messages = filter_messages(
    messages, include_types=["human", "ai"], exclude_ids=["3"]
)
print("\nFiltered by types and IDs:", filtered_messages)

File: ch4/js/e-filter-messages.js
import {
  HumanMessage,
  SystemMessage,
  AIMessage,
  filterMessages,
} from '@langchain/core/messages';

const messages = [
  new SystemMessage({ content: 'you are a good assistant', id: '1' }),
  new HumanMessage({ content: 'example input', id: '2', name: 'example_user' }),
  new AIMessage({
    content: 'example output',
    id: '3',
    name: 'example_assistant',
  }),
  new HumanMessage({ content: 'real input', id: '4', name: 'bob' }),
  new AIMessage({ content: 'real output', id: '5', name: 'alice' }),
];

// Filter for human messages
const filterByHumanMessages = filterMessages(messages, {
  includeTypes: ['human'],
});
console.log(`Human messages: ${JSON.stringify(filterByHumanMessages)}`);

// Filter to exclude names
const filterByExcludedNames = filterMessages(messages, {
  excludeNames: ['example_user', 'example_assistant'],
});
console.log(
  `\nExcluding example names: ${JSON.stringify(filterByExcludedNames)}`
);

// Filter by types and IDs
const filterByTypesAndIDs = filterMessages(messages, {
  includeTypes: ['human', 'ai'],
  excludeIds: ['3'],
});
console.log(
  `\nFiltered by types and IDs: ${JSON.stringify(filterByTypesAndIDs)}`
);
